src/nc_sentinel.c

```diff
@@ -0,0 +1,467 @@
+    ASSERT(!conn->client && !conn->proxy);
+    ASSERT(conn->status == CONN_DISCONNECTED);
+
+    return conn;
+}
+
+static void
+sentinel_set_next_connect(struct server_pool *pool)
+{
+    int64_t now;
+
+    now = nc_usec_now();
+    if (now < 0) {
+        /* get time failed, we reconnect immediately */
+        pool->next_sentinel_connect = 1LL;
+        return;
+    }
+
+    pool->next_sentinel_connect = now + pool->server_retry_timeout;
+}
+
+rstatus_t
+sentinel_connect(struct context *ctx, struct server *sentinel)
+{
+    rstatus_t status;
+    struct conn *conn;
+    struct msg *msg;
+    int cmd_num;
+    int i;
+
+    ASSERT(sentinel->sentinel);
+
+    /* get the only connect of sentinel */
+    conn = sentinel_conn(sentinel);
+    if (conn == NULL) {
+        /* can't call sentinel_close, manual set next connect */
+        sentinel_set_next_connect(sentinel->owner);
+        return NC_ENOMEM;
+    }
+
+    status = server_connect(ctx, sentinel, conn);
+    if(status != NC_OK) {
+        sentinel_close(ctx, conn);
+        return status;
+    }
+
+    /* set keepalive opt on sentinel socket to detect socket dead */
+    status = nc_set_tcpkeepalive(conn->sd, SENTINEL_KEEP_INTERVAL);
+    if (status < 0) {
+        log_error("set keepalive on s %d for sentienl server failed: %s",
+                  conn->sd, strerror(errno));
+        sentinel_close(ctx, conn);
+        return status;
+    }
+
+    cmd_num = sizeof(sentinel_reqs) / sizeof(char *);
+    for (i = 0; i < cmd_num; i++) {
+        msg = req_fake(ctx, conn);
+        if(msg == NULL) {
+            conn->err = errno;
+            sentinel_close(ctx, conn);
+            return NC_ENOMEM;
+        }
+
+        status = msg_append(msg, (uint8_t *)(sentinel_reqs[i]), nc_strlen(sentinel_reqs[i]));
+        if (status != NC_OK) {
+            conn->err = errno;
+            sentinel_close(ctx, conn);
+            return status;
+        }
+    }
+
+    conn->status = CONN_SEND_REQ;
+
+    sentinel->owner->next_sentinel_connect = 0LL;
+
+    return NC_OK;
+}
+
+void
+sentinel_close(struct context *ctx, struct conn *conn)
+{
+    struct server_pool *pool;
+
+    pool = ((struct server*)conn->owner)->owner;
+
+    sentinel_set_next_connect(pool);
+
+    conn->status = CONN_DISCONNECTED;
+
+    server_close(ctx, conn);
+}
+
+static rstatus_t
+sentinel_proc_sentinel_info(struct context *ctx, struct server *sentinel, struct msg *msg)
+{
+    rstatus_t status;
+    int i, master_num, switch_num, server_port;
+    struct string server_name, server_ip, sentinel_masters_prefix, tmp_string;
+    struct server *server;
+    struct mbuf *line_buf;
+
+    string_init(&tmp_string);
+    string_init(&server_name);
+    string_init(&server_ip);
+    string_set_text(&sentinel_masters_prefix, "sentinel_masters");
+
+    line_buf = mbuf_get();
+    if (line_buf == NULL) {
+        goto error;
+    }
+
+    /* get sentinel master num at line 3 */
+    msg_read_line(msg, line_buf, 3);
```

Comment:
This will/does have issues if redis-sentinel adds new fields to the response, changes the order of the fields, or sends more lines than this expects. See https://github.com/twitter/twemproxy/issues/527 (wrong repo)

After upgrading from redis 2.x to 3.2.6 , I noticed this. I'll try to find a solution

Reply:
https://github.com/ifwe/twemproxy/commits/build-nutredis contains a fix for this and other issues -  feel free to use https://github.com/ifwe/twemproxy/commit/9820caa202ab12bab01891174b6f66dbf37b9490

I've tested this with redis 3.2, and to some extent 4.0 (but not for long periods of time)